File: backend-api/src/main/java/com/project/backend_api/service/ProfileService.java
package com.project.backend_api.service;

import com.project.backend_api.dto.UserProfileDto;
import com.project.backend_api.model.User;
import com.project.backend_api.repository.UserRepository;
import com.project.backend_api.model.Gender;
import com.project.backend_api.repository.GenderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ProfileService {

    private final UserRepository userRepository;
    private final GenderRepository genderRepository;
    private final PasswordEncoder passwordEncoder;

    public UserProfileDto getUserProfile(UUID userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        return UserProfileDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .firstSurname(user.getFirstSurname())
                .secondSurname(user.getSecondSurname())
                .phoneNumber(user.getPhoneNumber())
                .phoneExtension(user.getPhoneExtension())
                .address(user.getAddress())
                .country(user.getCountry())
                .department(user.getDepartment())
                .city(user.getCity())
                .genderId(user.getGender() != null ? user.getGender().getId() : null)
                .genderName(user.getGender() != null ? user.getGender().getName() : null)
                .dateOfBirth(user.getDateOfBirth())
                .age(user.getAge())
                .pendingEmail(user.getPendingEmail())
                .isSuperAdmin(user.getIsSuperAdmin())
                .verified(user.getVerified())
                .build();
    }

    @Transactional
    public void updateProfile(UUID userId, UserProfileDto dto) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        user.setPhoneNumber(dto.phoneNumber());
        user.setPhoneExtension(dto.phoneExtension());
        user.setAddress(dto.address());
        user.setCountry(dto.country());
        user.setDepartment(dto.department());
        user.setCity(dto.city());
        user.setDateOfBirth(dto.dateOfBirth());
        user.setAge(dto.age());

        if (dto.genderId() != null) {
            Gender gender = genderRepository.findById(dto.genderId())
                    .orElseThrow(() -> new RuntimeException("Género no encontrado"));
            user.setGender(gender);
        }

        userRepository.save(user);
    }

    public boolean verifyPassword(UUID userId, String password) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));
        return passwordEncoder.matches(password, user.getPassword());
    }

    @Transactional
    public void changePassword(UUID userId, String oldPassword, String newPassword, String confirmPassword) {
        if (!newPassword.equals(confirmPassword)) {
            throw new RuntimeException("Las contraseñas no coinciden");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        if (!passwordEncoder.matches(oldPassword, user.getPassword())) {
            throw new RuntimeException("La contraseña actual es incorrecta");
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    @Transactional
    public void requestEmailChange(UUID userId, String newEmail) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        if (userRepository.findByEmail(newEmail).isPresent()) {
            throw new RuntimeException("El correo electrónico ya está en uso");
        }

        user.setPendingEmail(newEmail);
        user.setPendingEmailToken(UUID.randomUUID().toString());
        // In a real app, send verification email here
        userRepository.save(user);
    }
}

File: frontend-app/src/app/core/services/profile.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface UserProfile {
    id: string;
    username: string;
    email: string;
    firstName: string;
    firstSurname: string;
    secondSurname?: string;
    phoneNumber?: string;
    phoneExtension?: string;
    address?: string;
    country?: string;
    department?: string;
    city?: string;
    genderId?: string;
    genderName?: string;
    dateOfBirth?: string;
    age?: number;
    pendingEmail?: string;
    isSuperAdmin: boolean;
    verified: boolean;
}

@Injectable({ providedIn: 'root' })
export class ProfileService {
    private http = inject(HttpClient);
    private apiUrl = '/api/profile'; // Assuming this endpoint will be created/mapped

    getProfile(): Observable<UserProfile> {
        return this.http.get<UserProfile>(`${this.apiUrl}/me`);
    }

    updateProfile(data: Partial<UserProfile>): Observable<any> {
        return this.http.post(`${this.apiUrl}/update`, data);
    }

    verifyPassword(password: string): Observable<boolean> {
        return this.http.post<boolean>(`${this.apiUrl}/verify-password`, { password });
    }

    changeEmail(newEmail: string): Observable<any> {
        return this.http.post(`${this.apiUrl}/change-email`, { newEmail });
    }

    changePassword(oldPassword: string, newPassword: string, confirmPassword: string): Observable<any> {
        return this.http.post(`${this.apiUrl}/change-password`, { oldPassword, newPassword, confirmPassword });
    }

    getCountries(): Observable<any[]> {
        return this.http.get<any[]>('/api/geo/countries');
    }

    getStates(countryId: string): Observable<any[]> {
        return this.http.get<any[]>(`/api/geo/states?countryId=${countryId}`);
    }

    getCities(stateId: string): Observable<any[]> {
        return this.http.get<any[]>(`/api/geo/cities?stateId=${stateId}`);
    }

    getGenders(): Observable<any[]> {
        return this.http.get<any[]>('/api/geo/genders');
    }
}
